README.txt
Id Image Store
==============

Installation
------------

Dependencies:

* Python 2.4

* Python 2.4 compilation environment (including C compiler)

* pygame (optional)

To install::

  $ cd store

  $ python2.4 bootstrap.py
  
  $ bin/buildout

After this, the various scripts exist in the 'bin' directory.

Usage
-----

To start the store backend (on port 8080), run:

  $ bin/zopectl fg

You can access the web UI by browsing to
http://localhost:8080. Username 'admin', password 'admin'.

To create a new instance of the application, use 'Add application' and
name your application 'store' (or any name you wish, but the rest of
the examples below assume 'store').

To view the REST interface in your browser you can then browse to the
application (http://localhost:8080/store).

The filler script can be used to fill the store with data. It uses the REST
interface, so the backend needs to be running first.

If you use::

  $ bin/filler

then you will get a brief usage instruction.

Let's create a new session in the store, and create a group in that
session filling it with the images in a directory::

  $ bin/filler http://localhost:8080/store demo store /path/to/images_dir/

The filler script will now upload the images.

To see the changed state of the backend, you can look at
http://localhost:8080/store again.

To try the demo pygame frontend you need to have pygame installed. You can
then connect to the store with the following command:

  $ bin/frontend http://localhost:8080/store/sessions/demo

More information
----------------

Consult the following documents for technical information:

* src/imagestore/model.txt

* src/imagestore/rest.txt

* src/imagestore/search.txt

* src/imagestore/fake.txt

* src/imagestore/comprehensive.txt

There is also a website:

http://studiolab.io.tudelft.nl/imageSTORE/

License
-------

imageSTORE - a web service for the storing and organization of images

Copyright (C) 2007-2008, ID Studio Lab and Startifact

This library is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as
published by the Free Software Foundation; either version 2.1 of the
License, or (at your option) any later version.

This library is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA


CHANGES.txt
=======
CHANGES
=======

1.0b2 (unreleased)
==================

* Introduce an accounts and permissions infrastructure - see
  ``auth.txt`` for more information. In brief, there is an
  ``accounts`` structure under the store, and also a ``permissions``
  structure where accounts (include the ``default`` account, relevant
  for everybody) can be given the following permissions:

    - ``read`` (read-only)

    - ``write`` (read + write)

    - ``none`` (explicitly no permission)

  There is also a ``permissions`` structure for individual
  sessions. Users won't see or be allowed to access sessions that they
  do not at least have the ``read`` permission for.

  Permissions and accounts can be manipulated using RESTful apis.

1.0b1 (2008-07-29)
==================

* Verified that ``++rest++flash`` URL also shows up in absolute URLs.

* Added comprehensive tests for the HTTP methods (GET, POST, PUT,
  DELETE) for all kinds of resources. Make sure that the resources
  give back the right status codes.

* Resources that cannot be deleted now return the proper 405 status
  code and don't let themselves be deleted.

* Add LGPL license.

* Update to use Grok 0.13

* Update to use lxml 2.0.x

* remove ``time.sleep`` delay from 'filler' script.

0.9.2 (2008-06-03)
==================

* ++rest++flash form-based update system wasn't working properly with
  status messages. They weren't coerced to 200 always, and the real
  status information wasn't embedded in the returned XML.

0.9.1 (2008-06-02)
==================

* ++rest++flash URLs didn't work with the form upload functionality
  for images. This has been fixed.

0.9 (2008-05-22)
================

* Start maintaining this in (private) SVN instead of bzr.

* ++rest++flash in URLs now triggers Flash behavior.

* store image files as blobs. use ``z3c.blobfile`` as an backend,
  instead of ``zope.app.file``.

* add a special 'custom' metadata field wich can be used to store
  arbitrary XML.

* Upgrade to Grok 0.12

08-01-02
========

* Accidentally left debugging code in the system. Removed it now.

* The default behavior of the store now massively breaks the HTTP
  standard: no matter what status code should be returned, 200 is
  always returned. This is to please massively broken behavior of the
  Flash client, which apparently is unable to read any response body
  unless the status message is 200. To turn the behavior back to
  normal either override IHTTPOnly200 or disable IHTTPEmbed.

07-12-21
========

* Make sure content-type is correct for error responses.

* Embed HTTP info in 404 error response.

07-12-17
========

* XML responses for success and failure. Where previously a succesful
  response result in an empty body, success now gets an explicit
  ``<success />`` message. Errors previously just got a message as the
  body. They now use XML as well:
  ``<error><message>Reason</message></error>``

* HTTP information is embedded within the XML response. All responses
  (success or error) now have the HTTP response embedded in them. The
  first element directly under the root element will be an ``http``
  element. This element contains a ``status`` element containing HTTP
  status information, and zero or more ``header`` elements containing
  response headers.
 
* More XML is now copy & pasteable (GET the XML from server and then
  POST or PUT it back in works).

* Fix some failing tests introduced by the new ``href`` attribute on
  the ``source`` element.

07-12-13
========

* Allow faking PUT and DELETE over POST using special ``_method=put``
  and ``_method=delete`` URL parameters.

* You can POST to ``../images/factory`` now with a form-style
  (multipart/form-data) body, to create an image. This is not
  the official REST way to create images (POST the data to images),
  but systems like Flash cannot upload files any other way due to
  security reasons. Assumed is that the upload will take place using
  Flash's ``FileReference.upload()`` functionality, i.e. the form will
  have an entry called ``Filedata``.

* There is also a special URL ``../images/myimage/formdata`` that can
  be POSTed to using a form-style body to replace the image
  content. This is not the official REST way to replace images (PUT to
  the data to the image URL), but systems like Flash will be helped by
  this.

* Better error message if Slug header happens to be missing from the request.

* The 'source' element has its own href URL and can be modified using
  PUT. (it could already be, but its URL was missing and it was untested)

07-11-16
========

* rely on Grok 0.11 now, which has native REST support.

* A lot of renaming and reshuffling: renamed "session-library" to
  "sessions", renamed "image-library" to "images", renamed
  "library-image" to "source-image". "collection" element went
  away. Instead, the root group is called "collection". Make sure the
  filler script and frontend work with these changes.

* Sessions now have a new 'groups' sub-item that lists all the groups
  in the session.

* Added a test with non-ascii characters to verify whether XML coming
  from REST (or going into) can be properly UTF-8 encoded.

* No more i18n warning when starting up.

* A dashboard UI. To get to the dashboard UI, type '/dashboard' after
  the store root object.

* Extended the rest.txt doctest with a lot of text to clarify the procedure.

* Make sure that Location headers are set when necessary when doing a
  POST.

* Status strings are those as suggested by HTTP and more detailed
  status messages in case of errors are now in the response body.

* The Content-Type of the XML response is now ``application/xml``.

* Do not allow POST of object if object with the same name already exists
  in location: give 407 (Conflict) error instead.

* Fixed a bug where invalid XML could still result in the creation of
  an (invalid) object.

* various cleanups.

07-11-01
========

* Use Grok's REST infrastructure now. This cleans up things quite a
  bit.


trunk/src/imagestore/rest.py
import grok
from lxml import etree
import cgi

from zope import component
from zope.publisher.interfaces import INotFound

from imagestore.interfaces import IRest, IXml, IHTTPEmbed, IHTTPOnly200
from imagestore.xml import (export, Settings, NS,
                            XMLValidationError, XMLRecognitionError,
                            ContentNotAllowedError, CreationConflictError,
                            FactoryNotAllowedError, ReplaceNotAllowedError)

class StoreLayer(grok.IRESTLayer):
    pass

class FlashLayer(StoreLayer):
    pass

class Rest(grok.REST):
    grok.context(IRest)
    grok.layer(StoreLayer)
    
    def GET(self):
        self.response.setHeader('Content-Type',
                                'application/xml; charset=UTF-8')
        app_url = grok.url(self.request, grok.getSite())
        tree = export(self.context, Settings(expand='compact',
                                             app_url=app_url))
        self.response.setStatus(200)
        embed_http(self.request, self.response, tree)
        return etree.tostring(tree, encoding='UTF-8')

    def POST(self):
        self.response.setHeader('Content-Type',
                                'application/xml; charset=UTF-8')
        # fake PUT or DELETE if necessary
        msg = fakePutOrDelete(self.context, self.request)
        if msg is not None:
            return msg
        # normal POST behavior
        # set status in case what follows is successful. this
        # may overridden in the helper
        self.response.setStatus(201, 'Created')
        obj, tree = self._put_post_helper(
            lambda element: IXml(self.context).factory(element))
        # if we created a new object, or if there's an object already
        # there in conflict with ours, set the Location header with it
        if obj is not None:
            self.response.setHeader('Location', grok.url(self.request, obj))
        embed_http(self.request, self.response, tree)
        return etree.tostring(tree, encoding='UTF-8')
    
    def PUT(self):
        self.response.setHeader('Content-Type',
                                'application/xml; charset=UTF-8')
        self.response.setStatus(200, 'Ok')
        obj, tree = self._put_post_helper(
            lambda element: IXml(self.context).replace(element))
        embed_http(self.request, self.response, tree)
        return etree.tostring(tree, encoding='UTF-8')

    def _put_post_helper(self, func):
        try:
            element = etree.XML(self.body)
             # remove any http element that got in during roundtrip
            remove_http(element)
        except etree.XMLSyntaxError:
            self.response.setStatus(400, 'Bad Request')
            return None, error_message(
                "The submitted data was not well-formed XML.")
        try:
            return func(element), success_message()
        except FactoryNotAllowedError:
            self.response.setStatus(405, 'Method not allowed')
            return None, error_message(
                "Creation of content with POST is not allowed at this location.")
        except ReplaceNotAllowedError:
            self.response.setStatus(405, 'Method not allowed')
            return None, error_message(
                "Replacing of content with PUT is not allowed at this location.")
        except XMLRecognitionError:
            self.response.setStatus(400, 'Bad Request')
            return None, error_message(
                "The submitted XML could not be recognized.")
        except XMLValidationError:
            self.response.setStatus(400, 'Bad Request')
            return None, error_message("The submitted XML was invalid.")
        except ContentNotAllowedError:
            self.response.setStatus(400, 'Bad Request')
            return None, error_message(
                "The submitted content could not be added in this location.")
        except CreationConflictError, e:
            self.response.setStatus(409, 'Conflict')
            return (e.conflicting_object,
                    error_message("There is already a resource with this name "
                                  "in this location."))
        # shouldn't ever reach here
        assert False, "Failure in PUT or POST processing"

    def DELETE(self):
        self.response.setHeader('Content-Type',
                                'application/xml; charset=UTF-8')
 
        if IXml(self.context).is_deletable:
            self.response.setStatus(200, 'Ok')
            del self.context.__parent__[self.context.__name__]
            tree = success_message()
        else:
            self.response.setStatus(405, 'Method not allowed')
            tree = error_message(
                "This location may not be DELETEed")            
        embed_http(self.request, self.response, tree)
        return etree.tostring(tree, encoding='UTF-8')

class StoreProtocol(grok.RESTProtocol):
    grok.layer(StoreLayer)
    grok.name('store')

class FlashProtocol(StoreProtocol):
    grok.layer(FlashLayer)
    grok.name('flash')

def fakePutOrDelete(context, request):
    """Check whether to do fake PUT or DELETE.

    If fake needs to be performed, return message of fake request.
    If no faking is requested, return None.
    """
    qs = request.environment['QUERY_STRING']
    # this is an optimization so we don't have to do parse_qs all the time
    if not '_method=' in qs:
        return None
    d = cgi.parse_qs(qs)
    method = d.get('_method', None)
    # there actually was no _method parameter after all
    if method is None:
        return None
    method = method[0]
    method = method.lower()
    if method == 'put':
        view = component.getMultiAdapter((context, request),
                                         name='PUT')
        return view()
    elif method == 'delete':
        view = component.getMultiAdapter((context, request),
                                         name='DELETE')
        return view()
    # unknown fake HTTP method, ignore
    return None

class HTTPEmbed(grok.GlobalUtility):
    grok.provides(IHTTPEmbed)
    
    def __call__(self, request):
        return FlashLayer.providedBy(request)

class HTTPOnly200(grok.GlobalUtility):
    grok.provides(IHTTPOnly200)
    
    def __call__(self, request):
        return FlashLayer.providedBy(request)

def embed_http(request, response, tree):
    """Given an XML tree, embed HTTP response information in it.

    Also set actual response code to 200 if this turns out to be required
    by very broken clients.
    """
    is_http_embedded = component.getUtility(IHTTPEmbed)
    if not is_http_embedded(request):
        return
    http_el = etree.Element('{%s}http' % NS)
    root = tree.getroot()
    # move any text to after the HTTP element
    text = root.text
    # insert HTTP element
    root.insert(0, http_el)
    # make the text now the tail of HTTP
    http_el.tail = text
    # no more text
    root.text = None
    
    status_el = etree.SubElement(http_el, '{%s}status' % NS)
    status_el.attrib['code'] = str(response.getStatus())
    status_message = response.getStatusString()
    i = status_message.find(' ')
    status_message = status_message[i+1:]
    status_el.text = status_message
    for name, value in sorted(response.getHeaders()):
        if name == 'X-Powered-By':
            continue
        header_el = etree.SubElement(http_el, '{%s}header' % NS)
        header_el.attrib['name'] = name
        header_el.text = value

    # in very broken clients, only 200 response messages allow the
    # client to inspect the body. We force the response message back to
    # 200 at the last moment (after which we already exported the true
    # status message in the XML)
    response_only_200 = component.getUtility(IHTTPOnly200)
    if response_only_200(request):
        response.setStatus(200, 'Ok')

def remove_http(root):
    """Given root element, remove http element if it's there.
    """
    try:
        el = root[0]
    except IndexError:
        return
    if el.tag.endswith('}http'):
        del root[0]

def success_message():
    """Return XML success message.
    """
    success_el = etree.Element('{%s}success' % NS, nsmap={None:NS})
    tree = etree.ElementTree(element=success_el)
    return tree

def error_message(msg):
    """Return XML error message.
    """
    error_el = etree.Element('{%s}error' % NS,  nsmap={None:NS})
    message_el = etree.SubElement(error_el, '{%s}message' % NS)
    message_el.text = msg
    tree = etree.ElementTree(element=error_el)
    return tree

class NotFoundView(grok.View):
    grok.context(INotFound)
    grok.layer(StoreLayer)
    grok.name('index.html')

    def render(self):
        self.response.setHeader('Content-Type',
                                'application/xml; charset=UTF-8')
        self.response.setStatus(404)
        tree = error_message('Not found: %s' % self.request.getURL())
        embed_http(self.request, self.response, tree)
        return etree.tostring(tree, encoding='UTF-8')

